sw/device/examples/teacup_demos/teacup_leds_demo.c

```diff
@@ -0,0 +1,307 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_i2c.h"
+#include "sw/device/lib/dif/dif_spi_host.h"
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/i2c_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTTF_DEFINE_TEST_CONFIG();
+
+/**
+ * OT Peripheral Handles.
+ */
+static dif_i2c_t i2c;
+static dif_pinmux_t pinmux;
+
+/**
+ * Teacup RGB LEDs.
+ */
+typedef enum teacup_led {
+  kTeacupLedDs5 = 0,
+  kTeacupLedDs9 = 1,
+  kTeacupLedDs10 = 2,
+  kTeacupLedDs11 = 3,
+} teacup_led_t;
+
+/**
+ * LED RGB color value.
+ */
+typedef struct rgb_color {
+  uint8_t r;
+  uint8_t g;
+  uint8_t b;
+} rgb_color_t;
+
+/**
+ * Defined Constants.
+ */
+enum {
+  // Teacup LED Controller
+  kDeviceAddr = 0x34,
+
+  // Teacup LED Controller Registers
+  kControlRegAddr = 0x00,
+  kUpdateRegAddr = 0x49,
+  kLedChannel0ScalingRegAddr = 0x4D,
+  kGlobalCurrentControlRegAddr = 0x6E,
+
+  // LED DS5
+  // Blue
+  kPwmReg0LowAddr = 0x07,
+  // Red
+  kPwmReg1LowAddr = 0x09,
+  // Green
+  kPwmReg2LowAddr = 0x0B,
+
+  // LED DS9
+  // Blue
+  kPwmReg3LowAddr = 0x0D,
+  // Red
+  kPwmReg4LowAddr = 0x0F,
+  // Green
+  kPwmReg5LowAddr = 0x11,
+
+  // LED DS10
+  // Blue
+  kPwmReg6LowAddr = 0x13,
+  // Red
+  kPwmReg7LowAddr = 0x15,
+  // Green
+  kPwmReg8LowAddr = 0x17,
+
+  // LED DS11
+  // Blue
+  kPwmReg9LowAddr = 0x19,
+  // Red
+  kPwmReg10LowAddr = 0x1B,
+  // Green
+  kPwmReg11LowAddr = 0x1D,
+
+  // Other
+  kTotalNumLedsOnBoard = 4,
+  kNumLedCycles = 100,
+  kNumColorsInCycle = 4,
+  kLedCyclePauseMilliseconds = 200,
+  kLedBrightnessLowPercent = 5,
+  kLedBrightnessHighPercent = 40,
+  kLedBrightnessStepPercent = 5,
+  kDefaultTimeoutMicros = 1000,
+};
+
+static const rgb_color_t kColorBlue = {
+    .r = 0x33,
+    .g = 0x69,
+    .b = 0xE8,
+};
+
+static const rgb_color_t kColorRed = {
+    .r = 0xD5,
+    .g = 0x0F,
+    .b = 0x25,
+};
+
+static const rgb_color_t kColorYellow = {
+    .r = 0xEE,
+    .g = 0xB2,
+    .b = 0x11,
+};
+
+static const rgb_color_t kColorGreen = {
+    .r = 0x00,
+    .g = 0x99,
+    .b = 0x25,
+};
+
+static status_t peripheral_init(void) {
```

Comment:
It might be nice to move some of this LED control stuff into a testlib so other programs could more easily use it.

You can send that in a follow-up PR if you want.